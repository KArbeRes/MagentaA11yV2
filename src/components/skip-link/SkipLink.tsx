import { useKeyboardNavigation } from 'hooks/useKeyboardNavigation';
import React, { MutableRefObject, useEffect, useRef } from 'react';
import { useLocation } from 'react-router-dom';

import './skip-link.scss';

interface SkipLinkProps {
  mainContentRef: MutableRefObject<HTMLDivElement | null>;
  liveRegionTestId?: string;
}

const SkipLink: React.FC<SkipLinkProps> = ({
  mainContentRef,
  liveRegionTestId,
}) => {
  const location = useLocation();
  const isKeyboardNavigation = useKeyboardNavigation();
  const spanRef = useRef<HTMLDivElement | null>(null);
  const skipLinkRef = useRef<HTMLButtonElement | null>(null);

  const formatNavigationMessage = () => {
    const pathSegments = location.pathname.trim().split('/').filter(Boolean);

    if (pathSegments.length === 0) return '';

    const [category, component, section] = pathSegments;

    const formattedComponent = component?.replace(/-/g, ' ') || '';
    const formattedSection = section ? `${section} ` : '';

    return `Navigated to: ${category}${
      formattedComponent
        ? ` "${formattedComponent} ${formattedSection}page"`
        : ' page'
    }`.trim();
  };

  useEffect(() => {
    if (isKeyboardNavigation) {
      spanRef.current?.focus();
    }
  }, [location.pathname, isKeyboardNavigation]);

  return (
    <>
      <span ref={spanRef} tabIndex={-1} aria-labelledby="atomic-region"></span>
      <button
        className="skip-link"
        ref={skipLinkRef}
        onClick={() => {
          mainContentRef.current?.focus();
        }}>
        Skip to main content
      </button>
      <div
        id="atomic-region"
        data-testid={liveRegionTestId}
        aria-live="polite"
        aria-atomic="true">
        {formatNavigationMessage()}
      </div>
    </>
  );
};

export default SkipLink;
