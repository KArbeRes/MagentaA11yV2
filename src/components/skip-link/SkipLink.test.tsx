import React, { createRef } from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import SkipLink from './SkipLink';

describe('SkipLink Component', () => {
  test('renders the SkipLink button', () => {
    const mainContentRef = createRef<HTMLDivElement>();

    render(
      <MemoryRouter>
        <SkipLink
          mainContentRef={mainContentRef}
          isKeyboardNavigation={false}
        />
      </MemoryRouter>
    );

    const button = screen.getByRole('button', {
      name: /skip to main content/i,
    });
    expect(button).toBeInTheDocument();
  });

  test('focuses on the SkipLink button when isKeyboardNavigation is true', () => {
    const mainContentRef = createRef<HTMLDivElement>();

    render(
      <MemoryRouter>
        <SkipLink mainContentRef={mainContentRef} isKeyboardNavigation={true} />
      </MemoryRouter>
    );

    const button = screen.getByRole('button', {
      name: /skip to main content/i,
    });
    expect(button).toHaveFocus();
  });

  test('focuses on the main content when SkipLink button is clicked', () => {
    const mainContentRef = { current: document.createElement('div') };
    mainContentRef.current.setAttribute('tabindex', '-1');
    document.body.appendChild(mainContentRef.current);

    render(
      <MemoryRouter>
        <SkipLink
          mainContentRef={mainContentRef}
          isKeyboardNavigation={false}
        />
      </MemoryRouter>
    );

    const button = screen.getByRole('button', {
      name: /skip to main content/i,
    });

    fireEvent.click(button);

    expect(mainContentRef.current).toHaveFocus();

    document.body.removeChild(mainContentRef.current);
  });

  test('does not focus the SkipLink button when isKeyboardNavigation is false', () => {
    const mainContentRef = createRef<HTMLDivElement>();

    render(
      <MemoryRouter>
        <SkipLink
          mainContentRef={mainContentRef}
          isKeyboardNavigation={false}
        />
      </MemoryRouter>
    );

    const button = screen.getByRole('button', {
      name: /skip to main content/i,
    });

    expect(button).not.toHaveFocus();
  });

  test('assigns the SkipLink button ref correctly', () => {
    const mainContentRef = createRef<HTMLDivElement>();

    render(
      <MemoryRouter>
        <SkipLink
          mainContentRef={mainContentRef}
          isKeyboardNavigation={false}
        />
      </MemoryRouter>
    );

    const button = screen.getByRole('button', {
      name: /skip to main content/i,
    });

    expect(button).toBeInstanceOf(HTMLButtonElement);
    expect(button).toBeTruthy();
  });

  test('clicking SkipLink does nothing if mainContentRef is null', () => {
    const mainContentRef = { current: null };

    render(
      <MemoryRouter>
        <SkipLink
          mainContentRef={mainContentRef}
          isKeyboardNavigation={false}
        />
      </MemoryRouter>
    );

    const button = screen.getByRole('button', {
      name: /skip to main content/i,
    });

    fireEvent.click(button);

    expect(button).not.toHaveFocus();
  });
});
