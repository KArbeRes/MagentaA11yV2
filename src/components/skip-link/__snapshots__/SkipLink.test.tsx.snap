// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkipLink Component SkipLink Component - Snapshot Test matches the snapshot 1`] = `
<DocumentFragment>
  <span
    aria-labelledby="atomic-region"
    tabindex="-1"
  />
  <button
    class="skip-link"
  >
    Skip to main content
  </button>
  <div
    aria-atomic="true"
    aria-live="polite"
    id="atomic-region"
  />
</DocumentFragment>
`;
